kp_qsc.go

```diff
@@ -0,0 +1,257 @@
+		}
+	}
+
+	// set same headers to easycurl
+	easy.Setopt(curl.OPT_HTTPHEADER, headers)
+
+	// set Session Timeout
+	easy.Setopt(curl.OPT_TIMEOUT, int(c.Config.Timeout))
+
+	// set Method, Default is GET
+	if req.Method == "POST" {
+		easy.Setopt(curl.OPT_POST, true)
+	} else if req.Method == "DELETE" {
+		easy.Setopt(curl.OPT_CUSTOMREQUEST, "DELETE")
+	} else if req.Method == "PATCH" {
+		easy.Setopt(curl.OPT_CUSTOMREQUEST, "PATCH")
+	} else if req.Method == "PUT" {
+		easy.Setopt(curl.OPT_CUSTOMREQUEST, "PUT")
+	}
+	var curlresponse bytes.Buffer
+	retryCount := 1
+	dataTransferred := 0
+	// Note this func is invovked multiple times if data is larger than
+	// max size (16K). This functional always return true and errors are handled by checking data transferred.
+	// This is because of the way go-curl supports it.
+	// TODO: fix go-curl to return error on false. Currently even when false is returned, it returns success.
+	easy.Setopt(curl.OPT_WRITEFUNCTION,
+		func(ptr []byte, userdata interface{}) bool {
+			dataSize := len(string(ptr))
+			dataTransferred += dataSize
+
+			if curlresponse.Cap() < (curlresponse.Len() + dataSize) {
+				curlresponse.Grow(dataSize)
+			}
+			writtenSize, err := curlresponse.Write(ptr)
+			if err != nil {
+				c.Logger.Info(" curl writefunc error writing data ", err.Error())
+				curlresponse.Reset()
+			}
+
+			if writtenSize != dataSize {
+				c.Logger.Info(" curl writefunc cannot write full response", writtenSize, dataSize, curlresponse.Len())
+				curlresponse.Reset()
+			}
+
+			return true
+		})
+
+	if req.Body != nil {
+		b, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			c.Logger.Info(err.Error())
+			return nil, err
+		}
+		easy.Setopt(curl.OPT_POSTFIELDS, string(b))
+		// Need to write Body back
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+	}
+	responseHdrs := map[string]string{}
+	easy.Setopt(curl.OPT_HEADERFUNCTION,
+		func(ptr []byte, userdata interface{}) bool {
+			result := strings.Split(string(ptr), ":")
+
+			if len(result) > 1 {
+				for i := range result {
+					responseHdrs[result[i]] = result[i+1]
+					break
+				}
+			}
+			return true
+		})
+	curlstatusCode := 0
+
+	for {
+		err := easy.Perform()
+		if err != nil {
+			if strings.Contains(err.Error(), "Timeout") {
+				// retry
+			} else {
+				c.Logger.Info("CURL perform error", err.Error())
+				return nil, err
+			}
+		}
+
+		if dataTransferred != curlresponse.Len() {
+			return nil, errors.New("response data transfer failed")
+		}
+
+		curlCode, err := easy.Getinfo(curl.INFO_RESPONSE_CODE)
+		if err != nil {
+			c.Logger.Info("CURL failure getting response code error: ", err.Error())
+			return nil, err
+		}
+		curlstatusCode = curlCode.(int)
+		//c.Logger.Info("CURL statusCode: ", curlstatusCode)
+		// Retry on connection errors, 500+ errors (except 501 - not implemented), and 429 - too many requests
+		if !(curlstatusCode == 0 || curlstatusCode == 429 || (curlstatusCode >= 500 && curlstatusCode != 501)) {
```

Comment:
```suggestion
		if curlstatusCode != 0 && curlstatusCode != http.StatusTooManyRequests && (curlstatusCode < http. StatusInternalServerError || curlstatusCode == http. StatusNotImplemented)) {
```

Reply:
personally I find this a little easier to read :P

Reply:
done